In [1]:
import yfinance as yf
import numpy as np
import datetime as dt
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt

In [2]:
def calculate_annual_return(closing_prices):
    daily_returns = np.diff(closing_prices) / closing_prices[:-1]
    print(daily_returns)
    average_daily_return = np.mean(daily_returns)
    print(average_daily_return)
    trading_days = 252  # Assuming 252 trading days in a year
    annual_return = (1 + average_daily_return) ** trading_days - 1
    return annual_return

def calculate_return_volatility(closing_prices):
    # Calculate daily returns
    returns = np.diff(closing_prices) / closing_prices[:-1]
    #Calculate mean return
    mean_return = np.mean(returns)
    # Calculate variance of returns (There is no NaN values in the numpy array)
    var_stock = ((returns - mean_return)**2).sum() / (len(returns) -1)
    std_stock = var_stock**.5
    
    return std_stock

In [3]:
# Ticker symbol for the stock
ticker_symbol = 'NVDA'

# Fetching historical data using yfinance
stock_data = yf.download(ticker_symbol, start='2023-10-01', end='2023-12-31')

# Extracting the closing prices from the stock data
closing_prices = stock_data['Close'].values

# Calculating annual return and return volatility
annual_return = calculate_annual_return(closing_prices)
return_volatility = calculate_return_volatility(closing_prices)
annual_return_volatility = return_volatility * (252**0.5)

[*********************100%***********************]  1 of 1 completed
[-0.02824794  0.01204125  0.01469086  0.02403328 -0.01068569  0.01159632
  0.02200966  0.00296974 -0.03161152  0.01394608 -0.04679468 -0.03964681
 -0.00225135 -0.01695925  0.03836955  0.01600932 -0.04314865 -0.03477823
  0.00431481  0.01632095 -0.00925633  0.03788625  0.02790312  0.034455
  0.01657599  0.00445887  0.0134697   0.0080732   0.02949948  0.00589636
  0.02130807 -0.01546639  0.01210928 -0.00367821  0.02253638 -0.00922453
 -0.02458754 -0.0192955   0.00975386 -0.00872688  0.00667071 -0.02845862
 -0.00010695 -0.02683628  0.02320369 -0.02282782  0.02402038  0.01952959
 -0.01850294  0.02209025  0.00904379  0.00544833  0.01116855  0.02427898
 -0.00944542 -0.03009843  0.01827027 -0.00326599  0.00919521  0.00280039
  0.00212475  0.        ]
0.0018414565147013744


c:\Users\bert0\OneDrive\文件\quantcc\env_ds_39\lib\site-packages\yfinance\base.py:302: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  df["Dividends"].fillna(0, inplace=True)
c:\Users\bert0\OneDrive\文件\quantcc\env_ds_39\lib\site-packages\yfinance\base.py:303: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always 

In [4]:
print("Annual Return:", annual_return)
print("Return Volatility:", return_volatility)
print("Annual Return Volatility:", annual_return_volatility)

Annual Return: 0.5898192082743212
Return Volatility: 0.02096583000082576
Annual Return Volatility: 0.3328222332734525


In [5]:
stock_data

Open        High         Low       Close   Adj Close  \
Date                                                                     
2023-10-02  440.299988  451.750000  438.609985  447.820007  447.759644   
2023-10-03  448.079987  451.299988  432.459991  435.170013  435.111328   
2023-10-04  437.420013  441.429993  432.920013  440.410004  440.350616   
2023-10-05  440.500000  449.000000  438.880005  446.880005  446.819763   
2023-10-06  441.929993  457.890015  440.260010  457.619995  457.558319   
...                ...         ...         ...         ...         ...   
2023-12-22  491.950012  493.829987  484.670013  488.299988  488.277069   
2023-12-26  489.679993  496.000000  489.600006  492.790009  492.766907   
2023-12-27  495.109985  496.799988  490.850006  494.170013  494.146820   
2023-12-28  496.429993  498.839996  494.119995  495.220001  495.196777   
2023-12-29  498.130005  499.970001  487.510010  495.220001  495.196777   

              Volume  
Date                  
2023-10-02  43329800  
2023-10-03  47085000  
2023-10-04  36182100  
2023-10-05  39348300  
2023-10-06  43339700  
...              ...  
2023-12-22  25213900  
2023-12-26  24420000  
2023-12-27  23364800  
2023-12-28  24658700  
2023-12-29  38869000  

[63 rows x 6 columns]

The formula to convert daily volatility to annual volatility is:
Annual Volatility = Daily Volatility * Number of Trading Days in a Year^0.5 
So, if you have the daily volatility as a percentage, you would multiply it by the square root of 252 to get the annual volatility.
For example, if the daily volatility is 1.5%, the annual volatility would be:
Annual Volatility = 1.5% × 252 ^ 0.5
Annual Volatility = 1.5% × 15.874
Annual Volatility ≈ 23.81 %